pkg/sql/idxrecommendations/idx_recommendations_cache.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package idxrecommendations
+
+import (
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/settings"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlstats"
+	"github.com/cockroachdb/cockroach/pkg/util/duration"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+)
+
+// indexRecKey is used to determine if we should generate an index recommendation.
+type indexRecKey struct {
+	stmtNoConstants string
+	database        string
+	planHash        uint64
+}
+
+type indexRecInfo struct {
+	lastGeneratedTs time.Time
+	recommendations []string
+	executionCount  int64
+}
+
+type IndexRecCache struct {
+	st *cluster.Settings
+
+	// uniqueIndexRecInfoLimit is the limit on number of unique index
+	// recommendations info we can store in memory.
+	uniqueIndexRecInfoLimit *settings.IntSetting
```

Comment:
generally settings are global variables -- it's pretty rare to want to have a struct that is parameterized over both the setting itself and the cluster settings object. A setting is really just a name and an offset.

Reply:
Done